Northwind/Northwind/Middleware/CachingMiddleware.cs

```diff
@@ -0,0 +1,98 @@
+using Microsoft.AspNetCore.Http;
+
+namespace Northwind.Middleware
+{
+    // You may need to install the Microsoft.AspNetCore.Http.Abstractions package into your project
+    public class CachingMiddleware
+    {
+        private readonly RequestDelegate next;
+        private readonly CachingOptions options;
+
+        public CachingMiddleware(RequestDelegate next, CachingOptions options)
+        {
+            this.next = next;
+            this.options = options;
+        }
+
+        public async Task Invoke(HttpContext httpContext)
+        {
+            try
+            {
+                var absolutePathCacheDirectory = Path.Combine(Environment.CurrentDirectory, options.CacheDirectory);
+                Directory.CreateDirectory(absolutePathCacheDirectory);
+                var pathToCachedFile = Path.Combine(absolutePathCacheDirectory,
+                    httpContext.Request.Path.Value.Remove(0, 1).Replace('/', Path.DirectorySeparatorChar) + ".cache");
+                // skip further processing if cache exists and not expired
+                if (File.Exists(pathToCachedFile))
+                {
+                    if (DateTime.Now.AddMinutes(-options.ExpirationTime) < File.GetCreationTime(pathToCachedFile))
+                    {
+                        using (var fileStream = File.OpenRead(pathToCachedFile))
+                        {
+                            fileStream.Seek(0, SeekOrigin.Begin);
+                            fileStream.CopyTo(httpContext.Response.Body);
+                            return;
+                        }
+                    }
+
+                    File.Delete(pathToCachedFile);
+                }
+
+                //proceed with usual request if cached image was not found or was cleared
+
+                var originalBodyStream = httpContext.Response.Body;
+                using (var responseBody = new MemoryStream())
+                {
+                    httpContext.Response.Body = responseBody;
+                    await next(httpContext);
+                    httpContext.Response.Body.Seek(0, SeekOrigin.Begin);
+                    if (httpContext.Response.ContentType == "image/bmp" &&
+                        Directory.GetFiles(absolutePathCacheDirectory, "*", SearchOption.AllDirectories).Length <
+                        options.MaxCacheSize)
+                    {
+                        Directory.CreateDirectory(Path.GetDirectoryName(pathToCachedFile));
+                        using (var fileStream = File.Create(pathToCachedFile))
+                        {
+                            httpContext.Response.Body.CopyTo(fileStream);
+                            httpContext.Response.Body.Seek(0, SeekOrigin.Begin);
+                        }
+                    }
+
+                    await responseBody.CopyToAsync(originalBodyStream);
+                }
+            }
+            catch (Exception)
+            {
+                await next(httpContext);
```

Comment:
I will think about it

Reply:
I quickly looked at this issue. I think something wrong with routing and possibly this env variable ASPNETCORE_ENVIRONMENT .

At least when I added this code in the exception handling 
    ...
            catch (Exception e)
            {
                var code = HttpStatusCode.InternalServerError; // 500 if unexpected
                var result = JsonConvert.SerializeObject(new { error = e.Message });
                httpContext.Response.ContentType = "application/json";
                httpContext.Response.StatusCode = (int)code;
                await httpContext.Response.WriteAsync(result);
            }
    ...
and left only this configuration in Startup.cs:

            app.UseExceptionHandler(new ExceptionHandlerOptions
            {
                ExceptionHandler = async context =>
                {
                    var exception = context.Features.Get<IExceptionHandlerFeature>().Error;

                    context.Response.Redirect("/Home/Error");
                }
            });

When I got the error in the caching middleware, then I saw that Home/Error also has been called. Some more investigation required :)